src/libAtomVM/otp_net.c

```diff
@@ -0,0 +1,322 @@
+#include <defaultatoms.h>
+#include <globalcontext.h>
+#include <interop.h>
+#include <nifs.h>
+#include <otp_net.h>
+#include <port.h>
+#include <term.h>
+
+#include <errno.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <unistd.h>
+
+// #define ENABLE_TRACE
+#include <trace.h>
+
+#define FAMILY_ATOM globalcontext_make_atom(global, ATOM_STR("\x6", "family"))
+#define PROTOCOL_ATOM globalcontext_make_atom(global, ATOM_STR("\x8", "protocol"))
+#define TYPE_ATOM globalcontext_make_atom(global, ATOM_STR("\x4", "type"))
+#define INET_ATOM globalcontext_make_atom(global, ATOM_STR("\x4", "inet"))
+#define ADDR_ATOM globalcontext_make_atom(global, ATOM_STR("\x4", "addr"))
+// NB. The erlang documentation states that an address info structure is a map
+// with an `address` key, but OTP-24 returns maps with an `addr` key.
+#define ADDRESS_ATOM globalcontext_make_atom(global, ATOM_STR("\x7", "address"))
+
+#define UNKNOWN_TABLE_VALUE -1
+
+static const AtomStringIntPair protocol_table[] = {
+    { ATOM_STR("\x3", "tcp"), IPPROTO_TCP },
+    { ATOM_STR("\x3", "udp"), IPPROTO_UDP },
+    SELECT_INT_DEFAULT(UNKNOWN_TABLE_VALUE)
+};
+
+static const AtomStringIntPair type_table[] = {
+    { ATOM_STR("\x5", "dgram"), SOCK_DGRAM },
+    { ATOM_STR("\x6", "stream"), SOCK_STREAM },
+    SELECT_INT_DEFAULT(UNKNOWN_TABLE_VALUE)
+};
+
+//
+// utilities
+//
+
+static term socket_tuple_from_addr(Context *ctx, uint32_t addr)
+{
+    term terms[4];
+    terms[0] = term_from_int32((addr >> 24) & 0xFF);
+    terms[1] = term_from_int32((addr >> 16) & 0xFF);
+    terms[2] = term_from_int32((addr >> 8) & 0xFF);
+    terms[3] = term_from_int32(addr & 0xFF);
+
+    return port_create_tuple_n(ctx, 4, terms);
+}
+
+static inline term make_error_tuple(term reason, Context *ctx)
+{
+    term error_tuple = term_alloc_tuple(2, &ctx->heap);
+    term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+    term_put_tuple_element(error_tuple, 1, reason);
+    return error_tuple;
+}
+
+static term eai_errno_to_term(int err, GlobalContext *glb)
+{
+    switch (err) {
+        case EAI_AGAIN:
+            return globalcontext_make_atom(glb, ATOM_STR("\x8", "eaiagain"));
+        case EAI_BADFLAGS:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaibadflags"));
+        case EAI_FAIL:
+            return globalcontext_make_atom(glb, ATOM_STR("\x7", "eaifail"));
+        case EAI_FAMILY:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaifamily"));
+        case EAI_MEMORY:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eaimemory"));
+        case EAI_NONAME:
+            return globalcontext_make_atom(glb, ATOM_STR("\x9", "eainoname"));
+        case EAI_SERVICE:
+            return globalcontext_make_atom(glb, ATOM_STR("\xA", "eaiservice"));
+        case EAI_SOCKTYPE:
+            return globalcontext_make_atom(glb, ATOM_STR("\xB", "eaisocktype"));
+
+            // These constants are not supported on ESP32
```

Comment:
So should we ifdef them?

Reply:
What platform-agnostic macro to use?

Reply:
I mean CMake could detect if these symbols exist and generate a define accordingly, like it's done with open flags.

Reply:
Have a look at the CMake changes